### DISCUSSION TITLE
YOLO Thread-Safe Inference - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#6208] [Docs]
# YOLO Thread-Safe Inference - Ultralytics YOLOv8 Docs

This guide provides best practices for performing thread-safe inference with YOLO models, ensuring reliable and concurrent predictions in multi-threaded applications.

https://docs.ultralytics.com/guides/yolo-thread-safe-inference/

**Comment by glenn-jocher** [comment]
ALL FYI we also implement threading lock in the YOLO Predictor class, so even unintentionally unsafe inference workflows with multiple models running in parallel should still be thread-safe, but you will get the best performance by instantiating models within threads as shown the examples above. 

Happy YOLOing!

**Reply by qwerdf4** [reply]
when i run the Thread-Safe Example, i got RuntimeError: can't register atexit after shutdown

**Reply by pderrenger** [reply]
@qwerdf4 it seems like you're encountering an issue where the `atexit` handlers are being registered after the Python interpreter has begun the shutdown process. This can happen if you're creating new threads or processes during the shutdown phase, or if the threads are not managed correctly and are still running when the main program is trying to exit.

To resolve this, ensure that all threads are properly joined (i.e., they have completed execution) before the main program exits. You can use the `thread.join()` method to wait for a thread to finish its work. Also, make sure to instantiate any YOLO models within the thread's run method or target function, not at the module level, to avoid such issues.

Here's a simplified example of how you might structure your code:

```python
from threading import Thread
from ultralytics import YOLO

def thread_safe_inference(model_path, source):
    model = YOLO(model_path)
    results = model.predict(source)
    # Handle results here

threads = []
for i in range(num_threads):
    # Create a new thread and pass the model path and source
    t = Thread(target=thread_safe_inference, args=('yolov8n.pt', 'data/images/'))
    t.start()
    threads.append(t)

# Wait for all threads to complete
for t in threads:
    t.join()
```

If you're still facing issues, it might be helpful to review the thread management in your application to ensure that threads are not being created during the interpreter shutdown phase. If the problem persists, consider providing a minimal reproducible example in the issue tracker, and we'll be glad to assist further. 🛠️

**Comment by qwerdf4** [comment]
when i run the Thread-Safe Example, i got RuntimeError: can't register atexit after shutdown

**Reply by pderrenger** [reply]
@qwerdf4 it seems like you're encountering an issue where the Python `atexit` module, which handles cleanup actions at program termination, is being called after the Python interpreter has initiated its shutdown process. This can happen if you're trying to register an exit function during or after the interpreter shutdown, particularly in multi-threaded environments where threads are not properly managed.

To resolve this, ensure that you're not creating new threads or performing operations that require cleanup after the interpreter has started shutting down. Here are a few steps you can take:

1. Make sure that all threads are properly joined before the main thread completes execution.
2. Avoid starting new threads late in the lifecycle of your application, especially after initiating a shutdown.
3. If you're using a `with` statement for your YOLO model inference, ensure it's properly scoped and not extending beyond the lifetime of the main thread.

If you're still facing issues, it might be helpful to review the threading logic in your application to ensure that resources are being managed correctly. If you need further assistance, please provide more context or code snippets, and we'll do our best to help you out. 🛠️
